rs/registry/canister/tests/data_migration_test.rs

```diff
@@ -0,0 +1,325 @@
+                let decoded = SubnetListRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(NODE_REWARDS_TABLE_KEY) {
+                let decoded = NodeRewardsTable::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with("blessed_replica_versions") {
+                let decoded = BlessedReplicaVersions::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(API_BOUNDARY_NODE_RECORD_KEY_PREFIX) {
+                let decoded = ApiBoundaryNodeRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(NODE_OPERATOR_RECORD_KEY_PREFIX) {
+                let decoded = NodeOperatorRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(REPLICA_VERSION_KEY_PREFIX) {
+                let decoded = ReplicaVersionRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(HOSTOS_VERSION_KEY_PREFIX) {
+                let decoded = HostosVersionRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(SUBNET_RECORD_KEY_PREFIX) {
+                let decoded = SubnetRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else if key.starts_with(DATA_CENTER_KEY_PREFIX) {
+                let decoded = DataCenterRecord::decode(m.value.as_slice()).unwrap();
+                serde_json::to_value(&decoded).unwrap()
+            } else {
+                panic!("Missing mapping for key {}", key);
+            };
+
+            let value = fixup_ids(value);
+
+            (key.to_string(), value)
+        })
+        .collect();
+
+    serde_json::to_string_pretty(&mapped).unwrap()
+}
+
+fn fixup_ids(mut value: Value) -> Value {
+    if let Value::Array(arr) = &value {
+        // Try to convert the JSON array of numbers into a Vec<u8>
+        let byte_vec: Option<Vec<u8>> = arr
+            .iter()
+            // Check if all elements are numbers that can fit in a u8
+            .all(|v| v.is_u64() && v.as_u64().unwrap() <= 255)
+            .then(|| arr.iter().map(|v| v.as_u64().unwrap() as u8).collect());
+
+        if let Some(bytes) = byte_vec {
+            match Principal::try_from_slice(&bytes) {
+                Ok(p) => return Value::String(p.to_text()),
+                Err(_) => return Value::Array(arr.to_vec()),
+            }
+        }
+    }
+
+    if let Value::Object(map) = &mut value {
+        for value in map.values_mut() {
+            *value = fixup_ids(std::mem::take(value));
```

Comment:
Oh wow. This takes advantage of the highly non-obvious fact that serde_json::Value implements Default. This is tricky. I'm not sure, but it might be better if this function takes `&mut serde_json::Value` instead of consuming `serde_json::Value`...